src/github.com/matrix-org/dendrite/cmd/syncserver-integration-tests/main.go

```diff
@@ -171,36 +192,58 @@ func testSyncServer(input, want []string, since string) {
 	)
 	cmd.Stderr = os.Stderr
 	cmd.Stdout = os.Stderr
+	return cmd
+}
+
+func testSyncServer(input, want []string, since string) {
+	// Write the logs to kafka so the sync server has some data to work with.
+	exe.DeleteTopic(inputTopic)
+	if err := exe.CreateTopic(inputTopic); err != nil {
+		panic(err)
+	}
+	if err := exe.WriteToTopic(inputTopic, canonicalJSONInput(input)); err != nil {
+		panic(err)
+	}
+
+	cmd := prepareSyncServer()
 	if err := cmd.Start(); err != nil {
 		panic("failed to start sync server: " + err.Error())
 	}
 	done := make(chan error, 1)
 
-	// TODO: Waiting 1s is racey. Maybe keep hitting it until it doesn't get Connection Refused?
-	time.AfterFunc(1*time.Second, doSyncRequest(done, want, since))
+	// We need to wait for the sync server to:
+	// - have created the tables
+	// - be listening on the given port
+	// - have consumed the kafka logs
+	// before we begin hitting it with /sync requests. We don't get told when it has done
+	// all these things, so we just continually hit /sync until it returns the right bytes.
+	// We can't even wait for the first valid 200 OK response because it's possible to race
+	// with consuming the kafka logs (so the /sync response will be missing events and
+	// therefore fail the test).
+	go syncRequestUntilSuccess(done, want[0], since)
 
-	// wait for it to die or timeout
+	// wait for the sync server to exit or our test timeout to expire
 	go func() {
-		cmdErr := cmd.Wait()
-		if cmdErr != nil {
-			exitErr, ok := cmdErr.(*exec.ExitError)
-			if ok {
-				fmt.Println("\nSYNC SERVER ERROR: (", exitErr, ")")
-				fmt.Println("sync server failed to run. If failing with 'pq: password authentication failed for user' try:")
-				fmt.Println("    export PGHOST=/var/run/postgresql\n")
-			}
-		}
-		done <- cmdErr
+		done <- cmd.Wait()
 	}()
 	select {
 	case <-time.After(timeout):
+		if reqErr := getLastRequestError(); reqErr != nil {
+			fmt.Println("Last /sync request error:")
+			fmt.Println(reqErr)
+		}
+
 		if err := cmd.Process.Kill(); err != nil {
 			panic(err)
 		}
 		panic("dendrite-sync-api-server timed out")
 	case err := <-done:
```

Comment:
I feel like it would be clearer to `case err, ok := <- done` here and `ok == true` (or rather, just `ok`) instead of `err != nil` to distinguish between the "test succeeded" and "sync server died" cases.

Or even use different channels...

YMMV though